Add PIC support in the srv6 VPN scenario.

zice312963205 opened the pull request:
What I did
[HLD] [BGP PIC HLD](https://github.com/sonic-net/SONiC/pull/1493)

Why I did it
This PR implements the separation of nh_srv6-related information, which was originally stored in the nexthop, in the srv6-vpn scenario. It generates a new pic nexthop that contains only the next-hop forwarding information. Meanwhile, the original nexthop containing nh_srv6 is used as pic context information and indexed separately.
This is used to achieve fast switching of nexthop group members by identifying the corresponding members in the nexthop group based on route changes, and then performing operations on the members.

This PR is temporarily for review purposes. Documentation and test cases will be added later, and a complete PR will be resubmitted.

cscarpitta commented on zebra/rt_netlink.c:
```diff
@@ -2289,6 +2289,9 @@ ssize_t netlink_route_multipath_msg_encode(int cmd, struct zebra_dplane_ctx *ctx
 	/* Table corresponding to this route. */
 	table_id = dplane_ctx_get_table(ctx);
+	if (fpm)
+		table_id = dplane_ctx_get_vrf(ctx);
```
Why are we overriding the table ID with the VRF ID here? Isn't this a breaking change?

zice312963205 replied:
This part uses table_id as VRF lookup in fpmsyncd. This should be placed in the FRR patch of SONiC instead of being modified directly here. I have fixed it.

cscarpitta commented on zebra/main.c:
```diff
@@ -65,6 +65,8 @@ struct mgmt_be_client *mgmt_be_client;
 /* Route retain mode flag. */
 int retain_mode = 0;
 
+bool fpm_pic_nexthop = true;
```
What is this variable used for? It is set to true here and never changed. It seems that it is unnecessary and can be removed?

zice312963205 replied:
fixed

cscarpitta commented on zebra/zebra_dplane.c:
```diff
@@ -3522,10 +3540,20 @@ int dplane_ctx_route_init(struct zebra_dplane_ctx *ctx, enum dplane_op_e op,
 		ctx->u.rinfo.nhe.id = nhe->id;
 		ctx->u.rinfo.nhe.old_id = 0;
+		if (nhe->pic_nhe) {
+			/**/
```
Please fix the empty comment.

zice312963205 replied:
fixed

cscarpitta reviewed (commented):
Some comments.

The commit should be broken into multiple commits (bgpd, lib, zebra, ...)
https://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#commit-guidelines

ton31337 commented on lib/nexthop.c:
```diff
@@ -642,6 +642,9 @@ void nexthop_add_srv6_seg6(struct nexthop *nexthop, const struct in6_addr *segs,
 	for (i = 0; i < num_segs; i++)
 		memcpy(&nexthop->nh_srv6->seg6_segs->seg[i], &segs[i],
 		       sizeof(struct in6_addr));
+	if (segs_src) {
```
nit: Please drop `{}` for a single statement (everywhere in the code).

zice312963205 replied:
removed

ton31337 reviewed (commented):
Do we need a topotest, and a documentation a bit?

pguibert6WIND commented on zebra/zebra_nhg.c:
```diff
@@ -688,18 +689,26 @@ static void handle_recursive_depend(struct nhg_connected_tree_head *nhg_depends,
 
+static bool zebra_need_to_create_pic(struct nexthop *nh)
+{
+	if (nh && nh->nh_srv6 && !sid_zero(nh->nh_srv6->seg6_segs))
```
It should be better if the pic configuration was configurable.
by the way; does it make sense to apply this only for SRv6 ?
what about MPLS ? and what about regular IP routes?

zice312963205 replied:
Yes. Previously, we only used the PIC design for the SRv6 scenario. In fact, the PIC design is also used for MPLS and VXLAN scenarios. I will modify this logic to be more generic later. Thank you.

pguibert6WIND commented:
documentation (for outputs) and test (to demonstrate the efficiency) are obviously welcome

zice312963205 commented:
> documentation (for outputs) and test (to demonstrate the efficiency) are obviously welcome

Sorry for the late reply. I have been busy with another project recently. The documentation and topotests are being supplemented locally. In the next one to two weeks, I will add the documentation and topotests to this PR as a separate commit. Thank you.

ton31337 commented:
Why closing this and opening a new one? It's not acceptable, can we continue with this PR because it retains the history, comments, etc.?

zice312963205 commented:
> Why closing this and opening a new one? It's not acceptable, can we continue with this PR because it retains the history, comments, etc.?

Sure, because I switched the submission branch and removed some redundant code, and also did a rebase, the changes are quite significant. So, I have resubmitted the PR. I have now pushed the latest modifications to the current branch and reopened the PR. Please help review it. Thank you.

pguibert6WIND commented:
Hi @zice312963205 

in the last code, you remove the usage of NEXTHOP_GROUP_PIC_NON_RECURSIVE, because it is only set.
does that mean the cleaning is not done ?

```
+			SET_FLAG(pic_nhe->flags, NEXTHOP_GROUP_PIC_NON_RECURSIVE);
+#define NEXTHOP_GROUP_PIC_NON_RECURSIVE	    (1 << 11)
```

pguibert6WIND commented on zebra/zebra_rib.c:
```diff
@@ -736,6 +740,72 @@ void rib_uninstall_kernel(struct route_node *rn, struct route_entry *re)
+			zlog_debug("%s: pic_nhe %ul become invaild during route %pRN deleted, update pic_nh dependents %ul",
+				   __func__, picnhe->id, rn, rb_node_dep->nhe->id);
+		UNSET_FLAG(rb_node_dep->nhe->flags, NEXTHOP_GROUP_INSTALLED);
+		zebra_nhg_install_kernel(rb_node_dep->nhe, ZEBRA_ROUTE_MAX);
```
To illustrate, if 172.31.0.3 is removed, the dependent NHGIDs that use 172.31.0.3 will be removed.
The NHGID will be attempted to be removed, but because there is an other nexthop, the NHE will be installed, but with only the remaining nexthop. Right?
```
B>  192.0.2.9/32 [200/0] (74) (pic_nh 77) via 192.0.2.5 (vrf default) (recursive), label 500, weight 1, 00:00:08
  *                                         via 172.31.0.3, r1-eth1 (vrf default), label 16005/500, weight 1, 00:00:08
  *                                         via 172.31.2.4, r1-eth2 (vrf default), label 16005/500, weight 1, 00:00:08
r1# show nexthop-group rib 77
ID: 77 (zebra)
     RefCnt: 1
     Uptime: 00:00:13
     VRF: default(bad-value)
     Valid
     Depends: (68)
        via 192.0.2.5 (vrf default) (recursive), weight 1
           via 172.31.0.3, r1-eth1 (vrf default), weight 1
           via 172.31.2.4, r1-eth2 (vrf default), weight 1
```

eddieruan-alibaba replied:
It will be started from 172.31.0.3 withdraw event to zebra. PIC fix up could be kicked in to fix this NHG via removing the failed path. The NHG still belongs to 172.31.0.3 and 172.31.2.4.

Zebra will inform protocol process (bgp) who owns 192.0.2.9/32 to recalculate the paths. BGP will inform zebra for 192.0.2.9/32 via  172.31.2.4. This would create a new NHG for 172.31.2.4,  rebind prefix 192.0.2.9/32 to this new NHG, and remove old NHG for  172.31.0.3 and 172.31.2.4 if no one uses it.

Currently, we use NH address as part of hash key for NHG, so we can't reuse NHG.

pguibert6WIND replied:
My understanding is that ZEBRA is first aware of the IP unreachability event. In frrouting, this means a /32 route disappeared. This route stands for a nexthop that is looked up and contains the list of PIC NHE that use that IP.

At first sight, I was thinking it would work on the case where I use the IS-IS routing protocol (because the 192.0.2.5 is both a BGP endpoint and an IGP endpoint). But my case is slighly different, the nexthop is slightly different from the searched one, mainly because I have core nexthops, which are not yet handled by the lookup algorithm described above.

So finally, I understand that for kernel case, the backwalk algorithm as it is today can be used, but only on the case where BGP direct peering is used.

Saying that, this code is interesting, since it is the first brick that is able to efficiently handle switchover at zebra level, and not at protocol level. I think next steps for the kernel case will be to handle the PIC core case.

pguibert6WIND commented:
@zice312963205 , in reply to our discussion, I think a consistent solution should be found for linux kernel too. I think it is possible to handle the missing nexthop group hierarchy in the linux kernel, by doing a backwalk of all the  NHE that use that pic_nhe in the `zebra_update_pic_nhe()` function.

zice312963205 commented on zebra/zebra_rib.c:
```diff
@@ -736,6 +740,72 @@ void rib_uninstall_kernel(struct route_node *rn, struct route_entry *re)
+	switch (afi) {
+	case AFI_IP:
+		nh = nexthop_from_ipv4(&p->u.prefix4, NULL, zvrf_id(zvrf));
+
```
In this PR, it is indeed not possible to handle non-/32 routes. I have currently found a solution to handle all scenarios, which requires using RNH to find the actual next hop. I will be pushing this modification later.

pguibert6WIND replied:
thanks. keep me informed.

pguibert6WIND commented on zebra/main.c:
```diff
@@ -82,22 +84,22 @@ uint32_t rt_table_main_id = RT_TABLE_MAIN;
+				   { "vrfwnetns", no_argument, NULL, 'n' },
+				   { "nl-bufsize", required_argument, NULL, 's' },
+				   { "v6-rr-semantics", no_argument, NULL, OPTION_V6_RR_SEMANTICS },
 #endif /* HAVE_NETLINK */
```
is it really necessary to change the indentation?

zice312963205 replied:
fixed

eddieruan-alibaba commented:
> When PIC is used, it seems the PIC nexthop is not aware if it should be used for ipv4 or ipv6.
> This is the case for NHG 46, but not for NHG45.

I didn't get your comments. NHG would be aware it is via v4 NH or via V6 NH. But it would not be aware what type of prefixes would via it. It could be VPNv6 or VPNv6 both via the same NHG, which is v4 NH plus vpn label. 

In your example, PIC NHG is NHG 46, a.k.a NHG without any context. label NHG would be NHG 45, which contains label context. The purpose for PIC NHG is to allow NH information be sharable. This way, we could have a quick fix up when an NH is lost.

zice312963205 commented:
> @zice312963205 , in reply to our discussion, I think a consistent solution should be found for linux kernel too. I think it is possible to handle the missing nexthop group hierarchy in the linux kernel, by doing a backwalk of all the NHE that use that pic_nhe in the `zebra_update_pic_nhe()` function.

In the latest changes, I added a picnh_dependents structure. This structure can assist with reverse lookup in the zebra_update_pic_nhe() function to update all NHEs to the kernel. I will help complete this part of the development later.

pguibert6WIND commented:
I will have a look.

mjstapp reviewed (commented):
I think I understand the interest in improving FRR's handling of some situations, but I have two broad concerns about this approach. First, I don't want to encode one of the various overlay/underlay combinations into zebra data-structure changes, permanently. And second, I'm concerned that there may be some platform-specific aspects of path-handling and repair that are being pushed into FRR, and that would be a problem also. There are several possible ways to make improvements here, and I don't think we should take this path at this time.

zice312963205 commented:
> I think I understand the interest in improving FRR's handling of some situations, but I have two broad concerns about this approach. First, I don't want to encode one of the various overlay/underlay combinations into zebra data-structure changes, permanently. And second, I'm concerned that there may be some platform-specific aspects of path-handling and repair that are being pushed into FRR, and that would be a problem also. There are several possible ways to make improvements here, and I don't think we should take this path at this time.

Thank you for your comments. I would like to explain this modification:
First, the intent of this modification is not simply to differentiate between underlay and overlay and handle them differently. Rather, it aims to find a more efficient way to quickly notify the forwarding plane when the next-hop reachability of a route changes.
Therefore, the essence of this modification is to attempt to find other routes that use this route as a next-hop and quickly remove this next-hop when Zebra receives a route withdrawal (possibly a BGP route withdrawal, interface down, etc.).

However, in the current Zebra design, when routes like SRv6, MPLS, VXLAN are used as next hops, corresponding SID, label, VNI, etc., are used as next-hop keys to create different nexthop members. This results in the challenge that I cannot find the corresponding nexthop members solely through route changes:
For example, consider the following nexthops:
nhg 1: 1.1.1.1, label:10000, nhg_dependents 4
nhg 2: 1.1.1.1, label:10001, nhg_dependents 4
nhg 3: 2.2.2.2, label:20000, nhg_dependents 4
nhg 4 (group): 1/2/3
When a large number of routes (e.g., BGP routes) carry these three next-hops, Zebra forms many routes pointing to nhg 4:
191.0.0.1 --> nhg 4
191.0.0.2 --> nhg 4
191.0.0.3 --> nhg 4
191.0.0.4 --> nhg 4
...
At this point, when a route using 1.1.1.1 as the next-hop is deleted, the current Zebra process is to notify 1.1.1.1 (nht) to BGP, then BGP reselects the route and resends all the routes, modifying their association from nhg 4 to nhg 3. But if the number of routes is large, the route-by-route issuance can cause prolonged traffic interruption.
My modification attempts to address this situation by updating nhg 4 in the FIB before Zebra notifies BGP to reselect routes:
nhg 4 (group): 1/2/3 ---> nhg 4 (group): 3.
However, when I need to find the nexthops associated with the route 1.1.1.1 (nhg 1 and nhg 2), the labels used as keys in nh make it difficult to accurately find the corresponding nexthops. The goal of creating pic_nh is to separate these context details from the nexthop for independent, pure nexthop processing:
nhg 1: 1.1.1.1, label:10000 |   ----> pic_nh 10: 1.1.1.1, pic_dependents: 1,2
nhg 2: 1.1.1.1, label:10001 |   ----> pic_nh 10: 1.1.1.1, pic_dependents: 1,2
nhg 3: 2.2.2.2, label:20000 |   ----> pic_nh 20: 2.2.2.2, pic_dependents: 3

When 1.1.1.1 is deleted as a route, pic_nh 10 can be directly found, and based on "pic_dependents" in pic_nh 10, the actual nhg 1/2 used by routes can be found. Then, based on "nhg_dependents" 4, update nhg 4 in the FIB. This approach allows traffic to quickly converge onto the 2.2.2.2 direction before all routes are reissued, significantly reducing traffic interruption time.
In summary, this modification is not intended to differentiate overlay/underlay nor is it specific to certain dataplanes. Instead, it is a general method to address convergence issues in next-hop failure scenarios through internal Zebra processing. 
If have any questions, feel free to ping me.

zice312963205 commented:
ci:rerun

zice312963205 commented:
ci:rerun

zice312963205 commented:
ci:rerun

zice312963205 commented on bgpd/bgp_zebra.c:
```diff
@@ -1425,6 +1426,15 @@ static void bgp_zebra_announce_parse_nexthop(
 
+			if (mpinfo->peer->connection->status == Established &&
+			    mpinfo->peer->su_local->sa.sa_family == AF_INET6) {
+				memcpy(&api_nh->seg6_src,
```
fixed

zice312963205 commented on lib/nexthop.c:
```diff
@@ -888,6 +891,7 @@ void nexthop_copy_no_recurse(struct nexthop *copy,
 	memcpy(&copy->rmap_src, &nexthop->rmap_src, sizeof(nexthop->rmap_src));
 	memcpy(&copy->rmac, &nexthop->rmac, sizeof(nexthop->rmac));
 	copy->rparent = rparent;
+	copy->nh_encap.vni = nexthop->nh_encap.vni;
```
fixed

zice312963205 commented on lib/zclient.c:
```diff
@@ -2273,6 +2286,11 @@ int zapi_nexthop_from_nexthop(struct zapi_nexthop *znh,
 		}
+		if (!sid_zero_ipv6(&nh->nh_srv6->seg6_src)) {
+			SET_FLAG(znh->flags, ZAPI_NEXTHOP_FLAG_SEG6_SRC);
+			memcpy(&znh->seg6_src, &nh->nh_srv6->seg6_src,
```
fixed

zice312963205 commented:
ci:rerun

zice312963205 commented:
ci:rerun

zice312963205 commented:
ci:rerun

pguibert6WIND reviewed (approved):
This is ok to me.
PIC EDGE is a minimal brick that works for both srv6 and MPLS, by using only BGP as routing protocol.
This is a good starting point before handling other use cases.

cscarpitta reviewed (approved):
LGTM

donaldsharp commented:
As discussed in the maintainers channel w/ Phillippe:

<img width="932" alt="Screenshot 2025-05-20 at 4 33 02 PM" src="https://github.com/user-attachments/assets/34039147-8ae7-4c36-b5be-db831508c113" />
